### PR TITLE
Sphinx Recon_surf and CerebNet API's

### PR DESCRIPTION
**Pull request by engrosamaali91**
Continuation of remaining sphinx API's.

### CODE REVIEW

File: CerebNet/apply_warp.py [hunk 1]
```diff
@@ -14,21 +14,63 @@
...
+    img_data : ndarray
+        The image data to be saved.
+    save_path : str
+        The path (including file name) where the image will be saved.
+    header : nibabel.Nifti1Header
+        The header information for the NIfTI file.
+    affine : ndarray
+        The affine matrix for the NIfTI file.
+
+    Returns
```
Comment (dkuegler) [inline CerebNet/apply_warp.py]: If there is no return, just put no return section

File: CerebNet/apply_warp.py [hunk 2]
```diff
@@ -14,21 +14,63 @@
...
+    lbl_path : str
+        Path to the label image corresponding to the T1 image, to be warped similarly.
+    warp_path : str
+        Path to the warp field file used to warp the images.
+    result_path : str
+        Directory path where the warped and cropped images will be saved.
+    patch_size : tuple of int
+        The dimensions (height, width, depth) cropped images after warping.
+
+    Returns
```
Comment (dkuegler) [inline CerebNet/apply_warp.py]: Same

File: CerebNet/apply_warp.py [hunk 3]
```diff
@@ -14,21 +14,63 @@
...
+    None
+        This function returns nothing.
+    """
+
     img_out = nib.Nifti1Image(img_data, header=header, affine=affine)
     print(f"Saving {save_path}")
     nib.save(img_out, save_path)
 
 
 def store_warped_data(img_path, lbl_path, warp_path, result_path, patch_size):
```
Comment (dkuegler) [inline CerebNet/apply_warp.py]: This function should probably be renamed main

File: CerebNet/run_prediction.py [hunk 4]
```diff
@@ -94,6 +102,19 @@ def setup_options():
 
 
 def main(args):
+    """
+    Main function to run the inference based on the given command line arguments.
```
Comment (dkuegler) [inline CerebNet/run_prediction.py:106]: add "cerebelum subsegmentation"

also add a reference to the CerebNet paper (see README)

File: CerebNet/run_prediction.py [hunk 5]
```diff
@@ -94,6 +102,19 @@ def setup_options():
...
+
+    Parameters
+    ----------
+    args : argparse.Namespace
+        Command line arguments parsed by `argparse.ArgumentParser`.
+
+    Returns
+    -------
```
Comment (dkuegler) [inline CerebNet/run_prediction.py:115]: should not return None, int or str -> int= return code; str = error message.
return int=0 --> success

Comment (dkuegler) [reply CerebNet/run_prediction.py:115]: This pattern is something I want to implement throughout FastSurfer.
-> main() functions return 0 if successfull, and an error message (str) if unsuccessful.
In general, they should also not Raise Exceptions (but exception objects might be logged).

File: doc/conf.py [hunk 6]
```diff
@@ -21,6 +21,8 @@
 import os
 sys.path.append(os.path.dirname(__file__) + '/..') 
 sys.path.append(os.path.dirname(__file__) + '/../recon_surf') 
+# sys.path.append(os.path.dirname(__file__) + '/../CerebNet')
```
Comment (dkuegler) [inline doc/conf.py]: not needed, remove comment

File: recon_surf/N4_bias_correct.py [hunk 7]
```diff
@@ -15,18 +15,16 @@
 
 
 # IMPORTS
+import image_io as iio
```
Comment (dkuegler) [inline recon_surf/N4_bias_correct.py]: this should be imported last (it is a module in recon_surf)

File: recon_surf/N4_bias_correct.py [hunk 8]
```diff
@@ -15,18 +15,16 @@
...
-from typing import Optional, cast, Tuple
 import logging
-
 import SimpleITK as sitk
 import numpy as np
-from numpy import typing as npt
-
-import image_io as iio
 
+from pathlib import Path
```
Comment (dkuegler) [inline recon_surf/N4_bias_correct.py:22]: overall the ordering of imports is incorrect, the groups we had before were good, please only reorder within the groups.

group 1: python native modules
group 2: external modules, eg. torch, numpy
group 3: internal modules, e.g. here image_io

File: recon_surf/N4_bias_correct.py [hunk 9]
```diff
@@ -100,13 +98,13 @@
...
-    """Command line option parser.
+    """
+    Command line option parser.
 
     Returns
     -------
     options
-        object holding options
-
+        Object holding options.
```
Comment (dkuegler) [inline recon_surf/N4_bias_correct.py]: "Namesape object ..."

File: recon_surf/N4_bias_correct.py [hunk 10]
```diff
@@ -278,30 +276,30 @@ def normalize_wm_aseg(
         target_wm: float = 110.,
         target_bg: float = 3.
 ) -> sitk.Image:
-    """Normalize WM image [MISSING].
+    """
+    Normalize WM image [MISSING].
```
Comment (dkuegler) [inline recon_surf/N4_bias_correct.py]: MISSING -> so the white matter has a mean intensity of `target_wm` and the backgroud has intensity `target_bg`

File: recon_surf/N4_bias_correct.py [hunk 11]
```diff
@@ -334,13 +332,18 @@ def normalize_img(
     Parameters
     ----------
     itk_image : sitk.Image
+        Input image to be normalized.
     itk_mask : sitk.Image | None
+        [MISSING].
     source_intensity : Tuple[float, float]
+        Source intensity range.
```
Comment (dkuegler) [inline recon_surf/N4_bias_correct.py:345]: tuple lowercase (everywhere in typing)

File: recon_surf/N4_bias_correct.py [hunk 12]
```diff
@@ -334,13 +332,18 @@ def normalize_img(
     Parameters
     ----------
     itk_image : sitk.Image
+        Input image to be normalized.
     itk_mask : sitk.Image | None
+        [MISSING].
```
Comment (dkuegler) [inline recon_surf/N4_bias_correct.py]: Brain mask, voxels inside the mask are guaranteed to be > 0.
 "| None" is optional

File: recon_surf/N4_bias_correct.py [hunk 13]
```diff
@@ -475,11 +479,12 @@ def get_brain_centroid(itk_mask: sitk.Image) -> np.ndarray:
     Parameters
```
Comment (dkuegler) [inline recon_surf/N4_bias_correct.py:485]: Get the brain centroid from a binary image.

File: recon_surf/N4_bias_correct.py [hunk 14]
```diff
@@ -475,11 +479,12 @@ def get_brain_centroid(itk_mask: sitk.Image) -> np.ndarray:
...
     ----------
     itk_mask : sitk.Image
+        [MISSING].
```
Comment (dkuegler) [inline recon_surf/N4_bias_correct.py]: Binary image to compute the centroid of its labeled region.

File: recon_surf/fs_balabels.py [hunk 15]
```diff
@@ -74,13 +74,13 @@
...
-    """Command line option parser.
+    """
+    Command line option parser.
 
     Returns
     -------
     options
-        object holding options
-
+        Object holding options.
```
Comment (dkuegler) [inline recon_surf/fs_balabels.py]: Namespace

File: recon_surf/fs_balabels.py [hunk 16]
```diff
@@ -74,13 +74,13 @@
 
 
 def options_parse():
-    """Command line option parser.
+    """
+    Command line option parser.
```
Comment (dkuegler) [inline recon_surf/fs_balabels.py]: Create a command line interface and return command line options.

File: recon_surf/lta.py [hunk 17]
```diff
@@ -28,28 +28,28 @@ def writeLTA(
         dst_fname: str,
         dst_header: Dict
 ) -> None:
-    """Write linear transform array info to a .lta file.
+    """
+    Write linear transform array info to a .lta file.
```
Comment (dkuegler) [inline recon_surf/lta.py]: ... **an** .lta file.

File: recon_surf/map_surf_label.py [hunk 18]
```diff
@@ -59,13 +59,13 @@
 
 
 def options_parse():
-    """Command line option parser.
+    """
+    Command line option parser.
```
Comment (dkuegler) [inline recon_surf/map_surf_label.py]: see other files

File: recon_surf/paint_cc_into_pred.py [hunk 19]
```diff
@@ -42,13 +43,13 @@
 
 
 def argument_parse():
-    """Command line option parser.
+    """
```
Comment (dkuegler) [inline recon_surf/paint_cc_into_pred.py:46]: see other file

File: recon_surf/paint_cc_into_pred.py [hunk 20]
```diff
@@ -79,22 +80,22 @@ def argument_parse():
...
     aseg_cc : npt.ArrayLike
-        Aseg segmentation with CC
+        Aseg segmentation with CC.
 
     Returns
     -------
     pred
-        Prediction with added CC
-
+        Prediction with added CC.
```
Comment (dkuegler) [inline recon_surf/paint_cc_into_pred.py]: we should not use "pred"/"prediction" here change to asegdkt in this file and create a TODO for another student assistant to rename the file (paint_cc_into_asegdkt or similar) to rename the file and functions.

**Comment by engrosamaali91** [comment]
CerebNet remaining API's will be added in a new PR.

File: recon_surf/rewrite_mc_surface.py [hunk 21]
```diff
@@ -21,14 +21,13 @@
 
 
 def options_parse():
-    """Command line option parser.
+    """
```
Comment (dkuegler) [inline recon_surf/rewrite_mc_surface.py:24]: see other file

File: recon_surf/rewrite_mc_surface.py [hunk 22]
```diff
@@ -54,19 +53,19 @@ def options_parse():
 
 
 def resafe_surface(insurf: str, outsurf: str, pretess: str) -> None:
-    """Take path to insurf and rewrite it to outsurf thereby fixing vertex locs flag error.
+    """
+    Take path to insurf and rewrite it to outsurf thereby fixing vertex locs flag error.
 
-    (scannerRAS instead of surfaceRAS after marching cube)
+    (scannerRAS instead of surfaceRAS after marching cube).
```
Comment (dkuegler) [inline recon_surf/rewrite_mc_surface.py]: @m-reuter this line makes no sense to me. maybe you can propose how to fix this line.

File: recon_surf/N4_bias_correct.py (hunk 8 above)
Comment (dkuegler) [reply recon_surf/N4_bias_correct.py:22]: should be propagated to changes to all files (of this PR)

File: recon_surf/N4_bias_correct.py (hunk 9 above)
Comment (dkuegler) [reply recon_surf/N4_bias_correct.py]: this is a repeating comment

File: recon_surf/fs_balabels.py (hunk 16 above)
Comment (dkuegler) [reply recon_surf/fs_balabels.py]: this is a repeating comment

File: recon_surf/rotate_sphere.py [hunk 23]
```diff
@@ -19,11 +19,12 @@
 # IMPORTS
 import optparse
 import numpy as np
-from numpy import typing as npt
 import sys
 import nibabel.freesurfer.io as fs
 import align_points as align
 
+from numpy import typing as npt
```
Comment (dkuegler) [inline recon_surf/rotate_sphere.py]: ordering like I described in N4_bias_correct.py file

File: recon_surf/rotate_sphere.py [hunk 24]
```diff
@@ -66,13 +67,13 @@
 
 
 def options_parse():
-    """Command line option parser.
+    """
```
Comment (dkuegler) [inline recon_surf/rotate_sphere.py:71]: see fs_balabels.py

File: recon_surf/rotate_sphere.py [hunk 25]
```diff
@@ -119,13 +121,12 @@ def align_aparc_centroids(
...
     label_ids : npt.ArrayLike
-        Ids of the centroid to be aligned. Defaults to []
+        Ids of the centroid to be aligned. Defaults to [].
 
     Returns
     -------
     R
-        Rotation Matrix
-
+        Rotation Matrix.
```
Comment (dkuegler) [inline recon_surf/rotate_sphere.py:130]: return type missing npt.NDArray[float]

File: recon_surf/sample_parc.py [hunk 26]
```diff
@@ -20,13 +20,13 @@
...
 import sys
 import numpy as np
-from numpy import typing as npt
 import nibabel.freesurfer.io as fs
 import nibabel as nib
 from scipy import sparse
 from scipy.sparse.csgraph import connected_components
 from lapy import TriaMesh
 from smooth_aparc import smooth_aparc
+from numpy import typing as npt
```
Comment (dkuegler) [inline recon_surf/sample_parc.py]: ordering

File: recon_surf/sample_parc.py [hunk 27]
```diff
@@ -64,13 +64,13 @@
 
 
 def options_parse():
-    """Command line option parser.
+    """
```
Comment (dkuegler) [inline recon_surf/sample_parc.py:69]: fix

File: recon_surf/sample_parc.py [hunk 28]
```diff
@@ -101,7 +101,8 @@ def options_parse():
     return options
 
 def construct_adj_cluster(tria, annot):
-    """Adjacency matrix of edges from same annotation label only.
+    """
+    Adjacency matrix of edges from same annotation label only.
```
Comment (dkuegler) [inline recon_surf/sample_parc.py]: Compute ...

File: recon_surf/sample_parc.py [hunk 29]
```diff
@@ -261,7 +265,7 @@ def sample_img(surf, img, cortex=None, projmm=0.0, radius=None):
         Image to sample.
         If type is str, read image from file.
     cortex : np.ndarray | str
-        Filename of cortex label or np.array with cortex indices
+        Filename of cortex label or np.array with cortex indices.
     projmm : float
         Sample projmm mm along the surface vertex normals (default=0).
     radius : float [optional] | None
```
Comment (dkuegler) [inline recon_surf/sample_parc.py]: float, optional

File: recon_surf/sample_parc.py [hunk 30]
```diff
@@ -380,7 +386,7 @@ def sample_parc (surf, seg, imglut, surflut, outaparc, cortex=None, projmm=0.0,
     outaparc : str
         Filename for output surface parcellation.
     cortex : np.ndarray | str
-        Filename of cortex label or np.ndarray with cortex indices
+        Filename of cortex label or np.ndarray with cortex indices.
     projmm : float
         Sample projmm mm along the surface vertex normals (default=0).
     radius : float [optional] | None
```
Comment (dkuegler) [inline recon_surf/sample_parc.py]: see other function

File: recon_surf/smooth_aparc.py [hunk 31]
```diff
@@ -54,13 +54,13 @@
 
 
 def options_parse():
-    """Command line option parser.
+    """
+    Command line option parser.
```
Comment (dkuegler) [inline recon_surf/smooth_aparc.py]: update

File: recon_surf/smooth_aparc.py [hunk 32]
```diff
@@ -79,11 +79,13 @@ def options_parse():
 
 
 def get_adjM(trias: npt.NDArray, n: int):
```
Comment (dkuegler) [inline recon_surf/smooth_aparc.py]: trias: npt.NDArray[int]

File: recon_surf/smooth_aparc.py [hunk 33]
```diff
@@ -79,11 +79,13 @@ def options_parse():
...
-    """Create symmetric sparse adjacency matrix of triangle mesh.
+    """
+    Create symmetric sparse adjacency matrix of triangle mesh.
 
     Parameters
     ----------
     trias : np.ndarray shape (m, 3) int
```
Comment (dkuegler) [inline recon_surf/smooth_aparc.py]: trias: npt.NDArray[int](m, 3)

File: recon_surf/smooth_aparc.py [hunk 34]
```diff
@@ -79,11 +79,13 @@ def options_parse():
...
+        [MISSING].
```
Comment (dkuegler) [inline recon_surf/smooth_aparc.py]: Triangle mesh matrix.

File: recon_surf/smooth_aparc.py [hunk 35]
```diff
@@ -368,19 +370,19 @@ def smooth_aparc_files(
         incortexname: str,
```
Comment (dkuegler) [inline recon_surf/smooth_aparc.py:370]: should probably be main

File: recon_surf/smooth_aparc.py [hunk 36]
```diff
@@ -368,19 +370,19 @@ def smooth_aparc_files(
...
     inaparcname : str
-        Annotation filepath and name of source
+        Annotation filepath and name of source.
     incortexname : str
-        Label filepath and name of source
+        Label filepath and name of source.
     outaparcname : str
-        Suface filepath and name of destination
-
+        Suface filepath and name of destination.
```
Comment (dkuegler) [inline recon_surf/smooth_aparc.py]: Surface

File: recon_surf/smooth_aparc.py [hunk 37]
```diff
@@ -368,19 +370,19 @@ def smooth_aparc_files(
         incortexname: str,
         outaparcname: str
 ) -> None:
-    """Smoothes aparc.
+    """
+    Smoothes aparc.
```
Comment (dkuegler) [inline recon_surf/smooth_aparc.py]: Read files, smooth the aparc labels on the surface and save the smoothed labels.

File: recon_surf/smooth_aparc.py [hunk 38]
```diff
@@ -250,7 +251,8 @@ def mode_filter(
 
 
 def smooth_aparc(surf, labels, cortex = None):
-    """Smoothes aparc and fills holes.
+    """
+    Smoothes aparc and fills holes.
```
Comment (dkuegler) [inline recon_surf/smooth_aparc.py]: Smooth aparc label regions on the surface and fill holes.

File: recon_surf/spherically_project_wrapper.py [hunk 39]
```diff
@@ -14,20 +14,20 @@
...
-from subprocess import Popen, PIPE
 import shlex
 import argparse
 from typing import Any
+from subprocess import Popen, PIPE
 
 
 def setup_options():
-    """Command line option parser.
+    """
```
Comment (dkuegler) [inline recon_surf/spherically_project_wrapper.py:24]: update

File: recon_surf/spherically_project_wrapper.py [hunk 40]
```diff
@@ -78,22 +79,22 @@ def call(command: str, **kwargs: Any) -> int:
 
 
 def spherical_wrapper(command1: str, command2: str, **kwargs: Any) -> int:
-    """Run the first command. If it fails the fallback command is run as well.
+    """
+    Run the first command. If it fails the fallback command is run as well.
```
Comment (dkuegler) [inline recon_surf/spherically_project_wrapper.py]: ... is run instead.

**Review by dkuegler — CHANGES_REQUESTED** [review]
Couple of smaller changes.

File: recon_surf/rewrite_mc_surface.py (hunk 22 above)
Comment (m-reuter) [reply recon_surf/rewrite_mc_surface.py]: I have no idea, this was implemented 4 years ago by @LeHenschel  here 59f3959b1d3961334

File: CerebNet/apply_warp.py (hunk 3 above)
Comment (engrosamaali91) [reply CerebNet/apply_warp.py]: - Function "store_warped_data()" renamed to main().
- Function on Line 45 
- Function being called from Line 113

File: CerebNet/run_prediction.py (hunk 5 above)
Comment (engrosamaali91) [reply CerebNet/run_prediction.py:115]: Noted.

File: recon_surf/N4_bias_correct.py (hunk 8 above)
Comment (engrosamaali91) [reply recon_surf/N4_bias_correct.py:22]: > group 1: python native modules
> group 2: external modules, eg. torch, numpy
> group 3: internal modules, e.g. here image_io

Regrouped as follows: 
### Group 1: Python native modules
import argparse
import logging
import sys
from pathlib import Path
from typing import Optional, cast, Tuple

### Group 2: External modules
import SimpleITK as sitk
import numpy as np
from numpy import typing as npt

### Group 3: Internal modules
import image_io as iio

File: recon_surf/paint_cc_into_pred.py (hunk 20 above)
Comment (engrosamaali91) [reply recon_surf/paint_cc_into_pred.py]: Replaced instances of 'pred' with 'asegdkt'.
Added a TODO comment in last line.

File: recon_surf/fs_balabels.py [hunk 41]
```diff
@@ -74,13 +74,13 @@
 
 
 def options_parse():
-    """Command line option parser.
+    """
+    Create a command line interface and return command line options.
 
     Returns
     -------
     options
```
Comment (dkuegler) [inline recon_surf/fs_balabels.py]: missing Type. argparse.Namespace

File: recon_surf/lta.py [hunk 42]
```diff
@@ -28,28 +28,28 @@ def writeLTA(
         dst_fname: str,
         dst_header: Dict
 ) -> None:
-    """Write linear transform array info to a .lta file.
+    """
+    An .lta file.
```
Comment (dkuegler) [inline recon_surf/lta.py]: This should be a sentence.

File: recon_surf/map_surf_label.py [hunk 43]
```diff
@@ -59,13 +59,13 @@
 
 
 def options_parse():
-    """Command line option parser.
+    """
+    Create a command line interface and return command line options.
 
     Returns
     -------
     options
```
Comment (dkuegler) [inline recon_surf/map_surf_label.py]: typing -> : argparse.Namespace

File: recon_surf/map_surf_label.py [hunk 44]
```diff
@@ -99,30 +99,30 @@ def writeSurfLabel(
...
     surf : npt.NDArray
-        Surface coordinations
+        Surface coordinations.
 
     Raises
     ------
     ValueError
-        Label and values should have same sizes
-
+        Label and values should have same sizes.
```
Comment (dkuegler) [inline recon_surf/map_surf_label.py]: should say a condition when ValueError is raised, e.g. "if label and values are not the same size."

File: recon_surf/map_surf_label.py [hunk 45]
```diff
@@ -204,36 +204,37 @@ def mapSurfLabel(
...
     trg_values : np.ndarray
-        target values
+        Target values.
 
     Raises
     ------
     ValueError
-        Label and trg vertices should have same sizes
-
+        Label and trg vertices should have same sizes.
```
Comment (dkuegler) [inline recon_surf/map_surf_label.py]: if...

File: recon_surf/paint_cc_into_pred.py [hunk 46]
```diff
@@ -42,13 +43,13 @@
 
 
 def argument_parse():
-    """Command line option parser.
+    """
+    Create a command line interface and return command line options.
 
     Returns
     -------
     options
```
Comment (dkuegler) [inline recon_surf/paint_cc_into_pred.py]: typing

File: recon_surf/paint_cc_into_pred.py [hunk 47]
```diff
@@ -79,22 +80,22 @@ def argument_parse():
...
+    """
+    Paint corpus callosum segmentation into prediction.
 
     Note, that this function modifies the original array and does not create a copy.
 
     Parameters
     ----------
-    pred : npt.ArrayLike
-        Deep-learning prediction
+    asegdt : npt.ArrayLike
```
Comment (dkuegler) [inline recon_surf/paint_cc_into_pred.py]: asegdkt

File: recon_surf/paint_cc_into_pred.py (hunk 20 above)
Comment (dkuegler) [reply recon_surf/paint_cc_into_pred.py]: not "prediction", but "segmentation map"

File: recon_surf/rewrite_mc_surface.py (hunk 22 above)
Comment (dkuegler) [reply recon_surf/rewrite_mc_surface.py]: @dkuegler

File: recon_surf/rotate_sphere.py [hunk 48]
```diff
@@ -66,13 +68,13 @@
 
 
 def options_parse():
-    """Command line option parser.
+    """
+    Create a command line interface and return command line options.
 
     Returns
     -------
     options
```
Comment (dkuegler) [inline recon_surf/rotate_sphere.py]: typing

File: recon_surf/sample_parc.py [hunk 49]
```diff
@@ -64,13 +66,13 @@
 
 
 def options_parse():
-    """Command line option parser.
+    """
+    Create a command line interface and return command line options.
 
     Returns
     -------
     options
```
Comment (dkuegler) [inline recon_surf/sample_parc.py]: typing

File: recon_surf/spherically_project_wrapper.py [hunk 50]
```diff
@@ -14,20 +14,20 @@
...
 
 
 def setup_options():
-    """Command line option parser.
+    """
+    Create a command line interface and return command line options.
 
     Returns
     -------
     options
```
Comment (dkuegler) [inline recon_surf/spherically_project_wrapper.py]: typing

**Review by dkuegler — CHANGES_REQUESTED** [review]
some more small changes as discussed

File: recon_surf/paint_cc_into_pred.py [hunk 51]
```diff
@@ -79,22 +80,22 @@ def argument_parse():
 
 
 def paint_in_cc(pred: npt.ArrayLike, aseg_cc: npt.ArrayLike) -> npt.ArrayLike:
-    """Paint corpus callosum segmentation into prediction.
+    """
+    Paint corpus callosum segmentation into prediction.
```
Comment (dkuegler) [inline recon_surf/paint_cc_into_pred.py]: prediction

**Review by dkuegler — APPROVED** [review]
All changes approved.

**Pull request merged**